### PR TITLE
[release/9.0] Enable Central Package Management

### LINKED ISSUE #34638: Move to central package management
See [Central Package Management](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management) and https://github.com/dotnet/aspire/pull/5164/files#diff-5baf5f9e448ad54ab25a091adee0da05d4d228481c9200518fcb1b53a65d4156R8

Also enable [Package Source Mapping](https://learn.microsoft.com/en-us/nuget/consume-packages/package-source-mapping), see https://github.com/dotnet/aspire/blob/main/NuGet.config#L21-L38

### LINKED ISSUE #34649: Fix NuGetAudit warnings

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #34638
Fixes #34649

Also ports https://github.com/dotnet/efcore/pull/34470 and https://github.com/dotnet/efcore/pull/34576

Related future work:
- https://github.com/dotnet/efcore/issues/34637
- https://github.com/dotnet/efcore/issues/34485
- https://github.com/dotnet/efcore/issues/34654
- https://github.com/dotnet/efcore/issues/34655
- https://github.com/dotnet/efcore/issues/34656
- https://github.com/dotnet/efcore/issues/22138

### Description

Due to a recent change in NuGet restore now shows CVE warnings for transitive packages even if the actual version that will be used will be the current one provided by SDK. CPM ensures that all transitive dependencies are pinned to a warning-free version.

### Customer impact

Warnings on restore when EF packages are referenced.

### How found

Customer reported.

### Regression

No

### Testing

Tested manually.

### Risk

Low.

### CODE REVIEW

File: test/EFCore.Specification.Tests/EFCore.Specification.Tests.csproj [hunk 1]
```diff
@@ -56,12 +56,12 @@
...
-    <PackageReference Include="NetTopologySuite.IO.GeoJSON" Version="4.0.0" />
-    <PackageReference Include="xunit.assert" Version="$(XUnitVersion)" />
-    <PackageReference Include="xunit.core" Version="$(XUnitVersion)" />
-    <PackageReference Include="xunit.runner.visualstudio" Version="$(XUnitRunnerVisualstudioVersion)" />
-    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
+    <PackageReference Include="NetTopologySuite" />
+    <PackageReference Include="NetTopologySuite.IO.GeoJSON" />
+    <PackageReference Include="xunit.assert" VersionOverride="$(XUnitVersion)" />
+    <PackageReference Include="xunit.core" VersionOverride="$(XUnitVersion)" />
+    <PackageReference Include="xunit.runner.visualstudio" VersionOverride="$(XUnitRunnerVisualstudioVersion)" />
```
Comment (AndriySvyryd) [inline test/EFCore.Specification.Tests/EFCore.Specification.Tests.csproj:63]: These packages can't be managed centrally due to https://github.com/dotnet/arcade/issues/13798

**Comment by ericstj** [comment]
To mitigate product risk on a change like this, can you please share a diff of the nuspecs?

The way I normally do this is as follows:
```
git checkout hash-preceedingchange
build -pack /p:NuSpecOutputPath=artifacts\specs\beforechange
git checkout changes
git clean -fxd .
build -pack /p:NuSpecOutputPath=artifacts\specs\afterchange
```

Then I toss the files from before change in a scratch repo, commit, then copy over after change, commit, then share.  Here's an example: https://github.com/ericstj/diffs/commit/ee014f17bd44c877fab22034657ae38fbe1267ec

By diffing the nuspecs you can see exactly what changes in your product dependencies (assuming all you ship is nuget packages).  If you ship more than that you might consider other diffs to measure risk.

**Comment by AndriySvyryd** [comment]
Diff: https://github.com/dotnet/efcore/pull/34667/commits/280a1f173cd918223a12194b28f6343191cd174b

File: src/EFCore.Cosmos/Storage/Internal/ByteArrayConverter.cs
Comment (artl93) [inline src/EFCore.Cosmos/Storage/Internal/ByteArrayConverter.cs:1]: Is this change related to the rest? Hard to tell.

Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Storage/Internal/ByteArrayConverter.cs:1]: Yes, this is part of the https://github.com/dotnet/efcore/pull/34470 port due to the new Newtonsoft.Json

File: Directory.Packages.props [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    <PackageVersion Include="SQLitePCLRaw.bundle_e_sqlite3" Version="$(SQLitePCLRawVersion)" />
+
+    <!-- testing dependencies -->    
+    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp.Analyzer.Testing.XUnit" Version="$(MicrosoftCodeAnalysisTestingVersion)" />
+    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp.CodeFix.Testing.XUnit" Version="$(MicrosoftCodeAnalysisTestingVersion)" />
+
+    <!-- dependencies only used in tests -->    
+    <PackageVersion Include="Azure.Identity" Version="$(AzureIdentityVersion)" />
+    <PackageVersion Include="Azure.ResourceManager.CosmosDB" Version="$(AzureResourceManagerCosmosDBVersion)" />
+    <PackageVersion Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="8.0.2" />
```
Comment (ericstj) [inline Directory.Packages.props:63]: Is it intentional that we use the 8.0 version of this rather than live flow out of dotnet/aspnetcore?   Same goes for the other below.

File: Directory.Packages.props [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    <PackageVersion Include="Microsoft.AspNetCore.OData" Version="8.2.5" />
+    <PackageVersion Include="Microsoft.Extensions.Configuration" Version="$(MicrosoftExtensionsConfigurationVersion)" />
+    <PackageVersion Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="$(MicrosoftExtensionsConfigurationEnvironmentVariablesVersion)" />
+    <PackageVersion Include="Microsoft.Extensions.Configuration.Json" Version="$(MicrosoftExtensionsConfigurationJsonVersion)" />    
+    <PackageVersion Include="BenchmarkDotNet" Version="0.13.5" />
+    <PackageVersion Include="IdentityServer4.EntityFramework" Version="4.1.2" />
+    <PackageVersion Include="Grpc.AspNetCore" Version="2.59.0" />
+    <PackageVersion Include="NetTopologySuite.IO.GeoJSON" Version="4.0.0" />
+    <PackageVersion Include="Newtonsoft.Json" Version="13.0.3" />
```
Comment (ericstj) [inline Directory.Packages.props:72]: I learned that this update does carry breaking changes when discussing https://github.com/Azure/azure-cosmos-dotnet-v3/issues/4674, so be careful.  I think we should be OK if it's just a test using it.

File: eng/Version.Details.xml [hunk 4]
```diff
@@ -53,6 +53,14 @@
       <Uri>https://github.com/dotnet/runtime</Uri>
       <Sha>dec716d1c4651b70e4710e781e832a64be1e713b</Sha>
     </Dependency>
+    <Dependency Name="System.Text.Encodings.Web" Version="9.0.0-rc.2.24456.9">
```
Comment (ericstj) [inline eng/Version.Details.xml:56]: So we added these couple dependencies from runtime, but I wonder how many you should really add for completeness sake?

You probably transitively reference many other runtime packages.  By omitting them from CPM you end up with somewhat of an adhoc combination of different dependencies rather than a consistent baseline.  You could examine the project.assets.json files from your build to see that - I bet there's a mix of stuff from 6.0, 8.0, and 9.0.

The way I imagined folks would do CPM + transitive pinning is to list the entire dependency closure.  That way you control everything and have the ability to keep it all up to date.

File: Directory.Packages.props (hunk 3 above)
Comment (AndriySvyryd) [reply Directory.Packages.props:72]: It actually would be a breaking change for users. But it's in a major version and we can justify it.

Comment (AndriySvyryd) [reply Directory.Packages.props:72]: Filed https://github.com/dotnet/EntityFramework.Docs/issues/4804

**Comment by ericstj** [comment]
> Diff: [280a1f1](https://github.com/dotnet/efcore/commit/280a1f173cd918223a12194b28f6343191cd174b)

There's a lot changed in this diff - do you understand the differences and do they make sense to you?  Which of the packages here ship to customers - will those be OK with the dependency updates in RC2?

I see no effective changes to Microsoft.EntityFramework.Core and Microsoft.EntityFramework.Core.Abstractions which is good, just not sure which of the others to give scrutiny.

Less worried about tests - better to test latest.

**Comment by AndriySvyryd** [comment]
> There's a lot changed in this diff - do you understand the differences and do they make sense to you? Which of the packages here ship to customers - will those be OK with the dependency updates in RC2?

As far as I can see those are all needed to avoid NuGetAudit warnings

File: eng/Version.Details.xml (hunk 4 above)
Comment (AndriySvyryd) [reply eng/Version.Details.xml:56]: Sure, but in practice it shouldn't make a difference for the user

File: Directory.Packages.props [hunk 5]
```diff
@@ -0,0 +1,87 @@
+<Project>
+  <!-- Workaround https://github.com/dependabot/dependabot-core/issues/8490 -->
+  <!-- This file gets imported for out-of-tree test runs also where eng/Versions.props isn't
+       available -->
+  <Import Project="eng/Versions.props" Condition="'$(VersionPrefix)' == '' and Exists('eng/Versions.props')" />
```
Comment (roji) [inline Directory.Packages.props:5]: Mostly just curious, but is there a good reason to still keep actual version numbers in eng/Versions.props? Especially given that below we seem to have a mix, where some versions are specified inline here and others there.

Otherwise we could just have them here...

File: benchmark/Directory.Build.props [hunk 6]
```diff
@@ -6,10 +6,10 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="BenchmarkDotNet" Version="0.13.5" />
-    <PackageReference Include="xunit.assert" Version="$(XunitVersion)" />
-    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="$(MicrosoftExtensionsConfigurationJsonVersion)" />
-    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" Version="$(MicrosoftExtensionsConfigurationEnvironmentVariablesVersion)" />
+    <PackageReference Include="BenchmarkDotNet" />
+    <PackageReference Include="xunit.assert" VersionOverride="$(XUnitVersion)" />
```
Comment (roji) [inline benchmark/Directory.Build.props:10]: Just curious why this specific thing is needed, rather than just using the regular mechanism...

**Review by roji — APPROVED** [review]

File: Directory.Packages.props (hunk 3 above)
Comment (roji) [reply Directory.Packages.props]: Just noting that we're introducing various other intentional breaking changes to the Cosmos provider specifically, so this is a good time to do this.

**Comment by ericstj** [comment]
> As far as I can see those are all needed to avoid NuGetAudit warnings

I don't think so - the changes for audit were super minimal to product assemblies and only added a couple direct references to tests.   The diffs you're seeing are due to applying the versions you used directly in the repo to any where they might occur indirectly.  So a single reference in a test might find its way to changing a product depdendency.

Take this one: https://github.com/dotnet/efcore/commit/280a1f173cd918223a12194b28f6343191cd174b#diff-3c9435b86bee3d0caf074ebc9ba651f092ea4efcfcd8c6891521486a0a8f24a1R22

You mention in the package list `dependencies only used in tests` but based on the name of the package `Microsoft.EntityFramework.SqlServer` I think this is a product assembly.  Seems it is getting upgraded from 1.11.4 to 1.12.0.  That's what I meant by reviewing the diff - can you please go through that diff and especially for product packages make a comment on whether or not you think the lifted dependency adds risk to deliver in RC2?  I can't really make that call as I'm not familiar with EFCore.  Like I said if you just focus on the product packages it's probably not too much to do, but better to catch it now then be surprised later.

**Comment by AndriySvyryd** [comment]
> You mention in the package list dependencies only used in tests but based on the name of the package Microsoft.EntityFramework.SqlServer

I'll move these so they only apply to tests

File: Directory.Packages.props (hunk 5 above)
Comment (AndriySvyryd) [reply Directory.Packages.props:5]: Arcade/Darc still needs most of them in eng/Versions.props
Perhaps eventually they'll be unified

File: eng/Version.Details.xml (hunk 4 above)
Comment (AndriySvyryd) [reply eng/Version.Details.xml:56]: And I'd rather remove the ones added once NuGet supports ProvidedByPlatform

File: benchmark/Directory.Build.props (hunk 6 above)
Comment (AndriySvyryd) [reply benchmark/Directory.Build.props:10]: These packages can't be managed centrally due to dotnet/arcade#13798

**Comment by ericstj** [comment]
> I'll move these so they only apply to tests

Ok, but please do comb through an updated diff and make sure you're comfortable with any changes to the product packages.

**Comment by AndriySvyryd** [comment]
Updated diff: https://github.com/dotnet/efcore/commit/6a324996fe1a04cca68d40a980ac30a18f1820c7

**Comment by ericstj** [comment]
@AndriySvyryd -- looking for you and some other EF team member to review that diff and comment on the changes to ensure that everything in shipping packages looks good to go.  It's too much change for me to know how to review nor do I know which of your packages are shipping vs not.

**Comment by AndriySvyryd** [comment]
> @AndriySvyryd -- looking for you and some other EF team member to review that diff and comment on the changes to ensure that everything in shipping packages looks good to go

I reviewed them thoroughly and everything looks as expected. Will wait for @roji to give it another pass tomorrow

**Comment by roji** [comment]
Reviewed [the diff](https://github.com/dotnet/efcore/commit/6a324996fe1a04cca68d40a980ac30a18f1820c7#diff-d1003f1446b8ab6bb17396a36334a7d0cf83cd1f59780d9e8743e8e753370b1bR21), LGTM.,

* I'm leaving aside test changes (looking at product ones only)
* Everything in Microsoft.Extensions, System.Text.Json, System.Formats.Asn1 should indeed be updated to the latest 9.0.
* SQLitePCLRaw.core updating to 2.1.10 is intended, LGTM.
* Newtonsoft.Json updating to 13.0.3 for Cosmos LGTM.
* Microsoft.CodeAnalysis and Microsoft.Build.Framework LGTM.

File: src/EFCore.Tasks/EFCore.Tasks.csproj [hunk 7]
```diff
@@ -49,13 +49,13 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetFramework)' != 'net472'">
-    <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildFrameworkVersion)" PrivateAssets="all" ExcludeAssets="Runtime" IsImplicitlyDefined="true" />
-    <PackageReference Include="Microsoft.Build.Tasks.Core" Version="$(MicrosoftBuildTasksCoreVersion)" PrivateAssets="all" ExcludeAssets="Runtime" IsImplicitlyDefined="true" />
-    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" PrivateAssets="all" ExcludeAssets="Runtime" IsImplicitlyDefined="true" />
+    <PackageReference Include="Microsoft.Build.Framework" PrivateAssets="all" ExcludeAssets="Runtime" />
+    <PackageReference Include="Microsoft.Build.Tasks.Core" PrivateAssets="all" ExcludeAssets="Runtime" />
```
Comment (ericstj) [inline src/EFCore.Tasks/EFCore.Tasks.csproj:53]: I think you can actually remove this dependency entirely which will remove the need to update Asn1 for this case.  Referncing Tasks.Core (all the MSBuild task implementation types) is pretty unusual.
